src/sklearn_raster/ufunc.py

```diff
@@ -0,0 +1,243 @@
+
+        Parameters
+        ----------
+        func : callable
+            A function to apply to the flattened array. The function should accept one
+            array of shape (pixels, bands) and return one or more arrays of the same
+            shape.
+        skip_nodata : bool, default True
+            If True, NoData and NaN values will be removed before passing the array to
+            `func`. This can speed up processing of partially masked arrays, but may be
+            incompatible with functions that expect a consistent number of samples.
+        nodata_output : float or int, default np.nan
+            NoData pixels in the input will be replaced with this value in the output.
+            If the value does not fit the array dtype returned by `func`, an error will
+            be raised.
+        nan_fill : float or int, optional
+            If `skip_nodata=False`, any NaNs in the input array will be filled with this
+            value prior to calling `func` to avoid errors from functions that do not
+            support NaN inputs. If None, NaNs will not be filled.
+        ensure_min_samples : int, default 1
+            The minimum number of samples that should be passed to `func`. If the array
+            is fully masked and `skip_nodata=True`, dummy values (`nan_fill` or 0) will
+            be inserted to ensure this number of samples. This is necessary for
+            functions that require non-empty array inputs, like some scikit-learn
+            `predict` methods. No effect if the array contains enough valid pixels or if
+            `skip_nodata=False`.
+        allow_cast : bool, default False
+            If True and the `func` output dtype is incompatible with the chosen
+            `nodata_output` value, the output will be cast to the correct dtype.
+            Otherwise, an error will be raised.
+        **kwargs : dict
+            Additional keyword arguments passed to `func`.
+        """
+        # Fill NaNs in the input array if they're not being skipped
+        flat_array = self.flat_array if skip_nodata else self._fill_flat_nans(nan_fill)
+
+        # Only skip NoData if there's something to skip
+        if skip_nodata and self._num_masked > 0:
+            flat_result = self._skip_nodata_apply(
+                func,
+                flat_array=flat_array,
+                ensure_min_samples=ensure_min_samples,
+                nodata_output=nodata_output,
+                allow_cast=allow_cast,
+                nan_fill=nan_fill,
+                **kwargs,
+            )
+            # NoData is now pre-masked
+            mask_nodata = False
+        else:
+            flat_result = func(flat_array, **kwargs)
+            mask_nodata = self._num_masked > 0
+
+        @map_function_over_tuples
+        def _unflatten_and_mask(result: NDArray) -> NDArray:
+            """Unflattening result to (y, x, band) and mask NoData."""
+            output_shape = [*self.array.shape[:2], -1]
```

Comment:
I forgot to warn if `nodata_output` is found in the func result like we discussed in #41. I'll add that in here (I think), conditional on a `check_output_for_nodata` parameter.

Reply:
3cc69d6 adds this check to `_validate_nodata_output` (instead of here) so that it can be re-used whether `skip_nodata` is True or False.